src/txmempool.h

```diff
@@ -306,17 +307,41 @@ struct ancestor_score {};
 class CBlockPolicyEstimator;
 
 /** An inpoint - a combination of a transaction and an index n into its vin */
-class CInPoint
+struct CInPoint
 {
+    const CTransaction *const ptx;
+    const uint32_t n;
+    constexpr size_t DynamicMemoryUsage() const noexcept { return 0; }
+    const COutPoint& GetOutPoint() const noexcept { return ptx->vin[n].prevout; }
+};
+
+/* reference to a COutPoint that is either "free" or inside a CTransaction spending it
+ *
+ * must not outlive the COutPoint
+ */
+class CSpendRef
+{
+private:
+    union {
+    const COutPoint *const spent;
+    const CTransaction *const spender;
+    };
+    const uint32_t n;
+    const bool haveSpender;
 public:
-    const CTransaction* ptx;
-    uint32_t n;
-
-    CInPoint() { SetNull(); }
-    CInPoint(const CTransaction* ptxIn, uint32_t nIn) { ptx = ptxIn; n = nIn; }
-    void SetNull() { ptx = NULL; n = (uint32_t) -1; }
-    bool IsNull() const { return (ptx == NULL && n == (uint32_t) -1); }
-    size_t DynamicMemoryUsage() const { return 0; }
+    /* COutPoint in a spending CTransaction */
+    constexpr CSpendRef(const CInPoint& inpoint) noexcept : spender(inpoint.ptx), n(inpoint.n), haveSpender(true) {}
+    /* "free" COutPoint */
+    constexpr CSpendRef(const COutPoint& outpoint) noexcept : spent(&outpoint), n(0), haveSpender(false) {}
+    const COutPoint& GetOutPoint() const noexcept {
+        return haveSpender ? GetInPoint().GetOutPoint() : *spent;
+    }
+    constexpr CInPoint GetInPoint() const noexcept {
+        return haveSpender ? CInPoint{ spender, n } : throw std::logic_error("!haveSpender in CSpendRef::GetInPoint()");
```

Comment:
`noexcept` guarantees `std::terminate` occurs if an exception escapes the function; it's a hard assert that's allowed inside a `constexpr`.